File: cmd/dungeon/persist.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"unicode"
)

type persistMode int

const (
	persistNone persistMode = iota
	persistCreate
	persistReuse
	persistDiscard
)

func persistedContainerName(paths []string) (string, error) {
	cwd, err := os.Getwd()
	if err != nil {
		return "", err
	}
	absCwd, err := filepath.Abs(cwd)
	if err != nil {
		return "", err
	}

	hashInputs := []string{absCwd}
	for _, path := range paths {
		absPath, err := filepath.Abs(path)
		if err != nil {
			return "", err
		}
		hashInputs = append(hashInputs, absPath)
	}

	sum := sha256.Sum256([]byte(strings.Join(hashInputs, "\n")))
	shortHash := hex.EncodeToString(sum[:])[:8]

	base := sanitizeContainerBase(filepath.Base(absCwd))
	return fmt.Sprintf("dungeon-%s-%s", base, shortHash), nil
}

func sanitizeContainerBase(name string) string {
	var builder strings.Builder
	for _, char := range name {
		if unicode.IsLetter(char) || unicode.IsNumber(char) || char == '-' || char == '_' || char == '.' {
			builder.WriteRune(char)
			continue
		}
		builder.WriteRune('-')
	}
	cleaned := strings.Trim(builder.String(), "-")
	if cleaned == "" {
		return "project"
	}
	return cleaned
}

func podmanContainerExists(name string) (bool, error) {
	cmd := exec.Command("podman", "container", "exists", name)
	if err := cmd.Run(); err != nil {
		if exitErr := new(exec.ExitError); errors.As(err, &exitErr) {
			if exitErr.ExitCode() == 1 {
				return false, nil
			}
		}
		return false, err
	}
	return true, nil
}

func podmanContainerRunning(name string) (bool, error) {
	cmd := exec.Command("podman", "inspect", "-f", "{{.State.Running}}", name)
	output, err := cmd.Output()
	if err != nil {
		return false, err
	}
	return strings.TrimSpace(string(output)) == "true", nil
}

func startContainer(name string) error {
	cmd := exec.Command("podman", "start", name)
	return runPodmanCommand(cmd)
}

func execIntoContainer(name string) error {
	cmd := exec.Command("podman", "exec", "-it", name, "bash")
	return runPodmanCommand(cmd)
}

func ensureContainerSession(name string) error {
	running, err := podmanContainerRunning(name)
	if err != nil {
		return err
	}
	if !running {
		if err := startContainer(name); err != nil {
			return err
		}
	}
	return execIntoContainer(name)
}

func discardContainer(name string) error {
	cmd := exec.Command("podman", "rm", "-f", name)
	return runPodmanCommand(cmd)
}

func runPersistedSession(opts options, paths []string) error {
	exists, err := podmanContainerExists(opts.containerName)
	if err != nil {
		return err
	}
	if exists {
		return ensureContainerSession(opts.containerName)
	}
	cmd, err := buildPodmanCommand(opts, paths)
	if err != nil {
		return err
	}
	return runPodmanCommand(cmd)
}

func runPodmanCommand(cmd *exec.Cmd) error {
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

File: src/container/persist.rs
use std::{
    path::PathBuf,
    process::{Command, Stdio},
};

use sha2::{Digest, Sha256};

use crate::{container::podman::CommandSpec, error::AppError};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PersistMode {
    None,
    Create,
    Reuse,
    Discard,
}

pub fn persisted_container_name(paths: &[String]) -> Result<String, AppError> {
    let cwd = std::env::current_dir()?;
    let abs_cwd = cwd.canonicalize()?;

    let mut hash_inputs = vec![abs_cwd.to_string_lossy().to_string()];
    for path in paths {
        let abs_path = PathBuf::from(path).canonicalize()?;
        hash_inputs.push(abs_path.to_string_lossy().to_string());
    }

    let mut hasher = Sha256::new();
    hasher.update(hash_inputs.join("\n"));
    let hash = hasher.finalize();
    let short_hash = hex::encode(hash)[..8].to_string();

    let base = sanitize_container_base(
        abs_cwd
            .file_name()
            .and_then(|s| s.to_str())
            .unwrap_or("project"),
    );

    Ok(format!("dungeon-{}-{}", base, short_hash))
}

pub fn sanitize_container_base(name: &str) -> String {
    let mut cleaned = String::new();
    for ch in name.chars() {
        if ch.is_ascii_alphanumeric() || ch == '-' || ch == '_' || ch == '.' {
            cleaned.push(ch);
        } else {
            cleaned.push('-');
        }
    }
    let trimmed = cleaned.trim_matches('-').to_string();
    if trimmed.is_empty() {
        "project".to_string()
    } else {
        trimmed
    }
}

pub fn container_exists(name: &str) -> Result<bool, AppError> {
    let status = Command::new("podman")
        .arg("container")
        .arg("exists")
        .arg(name)
        .status()?;
    Ok(status.success())
}

pub fn container_running(name: &str) -> Result<bool, AppError> {
    let output = Command::new("podman")
        .arg("inspect")
        .arg("-f")
        .arg("{{.State.Running}}")
        .arg(name)
        .output()?;
    Ok(String::from_utf8_lossy(&output.stdout).trim() == "true")
}

pub fn start_container(name: &str) -> Result<(), AppError> {
    run_podman(&["start", name])
}

pub fn exec_into_container(name: &str) -> Result<(), AppError> {
    run_podman(&["exec", "-it", name, "bash"])
}

pub fn ensure_container_session(name: &str) -> Result<(), AppError> {
    let running = container_running(name)?;
    if !running {
        start_container(name)?;
    }
    exec_into_container(name)
}

pub fn discard_container(name: &str) -> Result<(), AppError> {
    run_podman(&["rm", "-f", name])
}

pub fn run_persisted_session(name: &str, spec: CommandSpec) -> Result<(), AppError> {
    if container_exists(name)? {
        return Err(AppError::message(format!(
            "ERROR: container \"{}\" already exists, use --persisted to connect",
            name
        )));
    }
    let mut cmd = Command::new(spec.program);
    cmd.args(spec.args);
    cmd.stdin(Stdio::inherit());
    cmd.stdout(Stdio::inherit());
    cmd.stderr(Stdio::inherit());
    let status = cmd.status()?;
    if status.success() {
        Ok(())
    } else {
        Err(AppError::Subprocess(
            status.code().unwrap_or(1),
            "podman exited with error".to_string(),
        ))
    }
}

fn run_podman(args: &[&str]) -> Result<(), AppError> {
    let mut cmd = Command::new("podman");
    cmd.args(args);
    cmd.stdin(Stdio::inherit());
    cmd.stdout(Stdio::inherit());
    cmd.stderr(Stdio::inherit());
    let status = cmd.status()?;
    if status.success() {
        Ok(())
    } else {
        Err(AppError::Subprocess(
            status.code().unwrap_or(1),
            "podman exited with error".to_string(),
        ))
    }
}
